crates/pep440-rs/src/version.rs

```diff
@@ -876,100 +879,110 @@ impl VersionSmall {
 
     #[inline]
     fn post(&self) -> Option<u64> {
-        let v = (self.repr >> 16) & 0xFF;
-        if v == 0 {
-            None
+        if self.suffix_kind() == VersionSmall::SUFFIX_POST {
+            Some(self.suffix_version())
         } else {
-            Some(v - 1)
+            None
         }
     }
 
     #[inline]
     fn set_post(&mut self, value: Option<u64>) -> bool {
-        if value.is_some() && (self.pre().is_some() || self.dev().is_some()) {
-            return false;
+        if self.pre().is_some() || self.dev().is_some() {
+            return value.is_none();
         }
         match value {
             None => {
-                self.repr &= !(0xFF << 16);
+                self.set_suffix_kind(VersionSmall::SUFFIX_NONE);
             }
             Some(number) => {
-                if number > 0b1111_1110 {
+                if number > VersionSmall::SUFFIX_MAX_VERSION {
                     return false;
                 }
-                self.repr &= !(0xFF << 16);
-                self.repr |= (number + 1) << 16;
+                self.set_suffix_kind(VersionSmall::SUFFIX_POST);
+                self.set_suffix_version(number);
             }
         }
         true
     }
 
     #[inline]
     fn pre(&self) -> Option<PreRelease> {
-        let v = (self.repr >> 8) & 0xFF;
-        if v == 0xFF {
-            return None;
+        let (kind, number) = (self.suffix_kind(), self.suffix_version());
+        if kind == VersionSmall::SUFFIX_PRE_ALPHA {
+            Some(PreRelease {
+                kind: PreReleaseKind::Alpha,
+                number,
+            })
+        } else if kind == VersionSmall::SUFFIX_PRE_BETA {
+            Some(PreRelease {
+                kind: PreReleaseKind::Beta,
+                number,
+            })
+        } else if kind == VersionSmall::SUFFIX_PRE_RC {
+            Some(PreRelease {
+                kind: PreReleaseKind::Rc,
+                number,
+            })
+        } else {
+            None
```

Comment:
Should `number` be 0 here? (Should we assert?)

Reply:
Nope! If the kind isn't a pre-release, then it could be something else with a non-zero number.